### PR TITLE
Add testing

### PR DESCRIPTION
**Pull request by vjeux**
- This brings in the flow test suite that contains a ton of JavaScript parsing edge cases
- This creates snapshot tests using the pretty printer for all of them
- If uncomment `RUN_AST_TESTS` line in `tests/run_specs.js`, it checks ast(pretty_print(x)) == ast(x). Right now, "178 failed, 197 passed, 375 of 377 total". So half of the tests are not passing, most of them are crashes and many of the rest are subtle issues.

### CODE REVIEW

**Comment by vjeux** [comment]
For the errors, adding

```
String.prototype.join = function(arr) {
  return arr.join(this);
}
```

seems to be making a lot go away. Is this a new thing inside of node 7? (I'm using 6.9 right now)

**Comment by jlongster** [comment]
Wow, this is awesome, thanks so much!

> Is this a new thing inside of node 7?

No, this is a difference with the old recast API and the new refactoring. Recast's old printer generated a `Lines` object which was its implementation of keeping track of string as they were generated. That object has a `join` method, and several places of the code haven't been converted to the new algorithm and still expect a `Lines` object to exist.

Now many of those places get a raw string (from the new algorithm) and are trying to join on it, which is why that seems to fix it (but it doesn't really :)). Many places are trivial to fix. If it does `thing.join(",")`, you just need to change it to be `join(",", thing)`. There are probably subtleties though; usually you want to make a few other tweaks when updating the old code.

You can search for where `group` and `line` are used to find several examples of using the new API.

This test suite is *awesome* and is a perfect way to track the refactoring. We just need to get all these tests to pass! I've been able to refactor a ton very quickly and I'm optimistic it shouldn't be hard to finish all the node types.

File: package.json [hunk 1]
```diff
@@ -11,6 +11,19 @@
     "recast": "^0.11.18"
   },
   "devDependencies": {
-    "glob": "^7.1.1"
+    "glob": "^7.1.1",
+    "jest": "^18.0.0",
+    "flow-parser": "^0.37.0"
```
Comment (jlongster) [inline package.json]: Any reason this is here? I don't see it required anywhere

**Comment by jlongster** [comment]
Going ahead and merging it, can't really look through all of it as it's so big but I think it's a good idea trying to get this test suite running. Thanks!
